grafanalib/core.py

```diff
@@ -1071,3 +1092,168 @@ def to_json_data(self):
+
+    def to_json_data(self):
+        return {
+            'text': self.text,
+            'value': self.value,
+        }
+
+
+@attr.s
+class Table(object):
+    """Generates Table panel json structure
+
+    Grafana doc on table: http://docs.grafana.org/reference/table_panel/
+
+    :param columns: table columns for Aggregations view
+    :param dataSource: Grafana datasource name
+    :param description: optional panel description
+    :param editable: defines if panel is editable via web interfaces
+    :param fontSize: defines value font size
+    :param height: defines panel height
+    :param hideTimeOverride: hides time overrides
+    :param id: panel id
+    :param links: additional web links
+    :param minSpan: minimum span number
+    :param pageSize: rows per page (None is unlimited)
+    :param scroll: scroll the table instead of displaying in full
+    :param showHeader: show the table header
+    :param span: defines the number of spans that will be used for panel
+    :param styles: defines formatting for each column
+    :param targets: list of metric requests for chosen datasource
+    :param title: panel title
+    :param transform: table style
+    :param transparent: defines if panel should be transparent
+    """
+
+    dataSource = attr.ib()
+    targets = attr.ib()
+    title = attr.ib()
+    columns = attr.ib(default=attr.Factory(list))
+    description = attr.ib(default=None)
+    editable = attr.ib(default=True, validator=instance_of(bool))
+    fontSize = attr.ib(default="100%")
+    height = attr.ib(default=None)
+    hideTimeOverride = attr.ib(default=False, validator=instance_of(bool))
+    id = attr.ib(default=None)
+    links = attr.ib(default=attr.Factory(list))
+    minSpan = attr.ib(default=None)
+    pageSize = attr.ib(default=None)
+    repeat = attr.ib(default=None)
+    scroll = attr.ib(default=True, validator=instance_of(bool))
+    showHeader = attr.ib(default=True, validator=instance_of(bool))
+    span = attr.ib(default=6)
+    sort = attr.ib(default=attr.Factory(ColumnSort), validator=instance_of(ColumnSort))
+    styles = attr.ib()
+
+    transform = attr.ib(default=COLUMNS_TRANSFORM)
+    transparent = attr.ib(default=False, validator=instance_of(bool))
+
+    @styles.default
+    def styles_default(self):
+        return [
+            ColumnStyle(
+                alias="Time",
+                pattern="time",
+                type="date",
+            ),
+            ColumnStyle(
+                pattern="/.*/",
+            ),
+        ]
+
+    def to_json_data(self):
+        styles = self.styles
+        for column in self.columns:
+            if column.style:
+                style = copy.deepcopy(column.style)
+                if style.pattern and style.pattern != column.text:
+                    raise ValueError("Column style pattern must match the column name if specified")
+                style.pattern = column.text
+                styles.append(style)
```

Comment:
I don't understand what this is trying to achieve. Could you perhaps provide an example of a `Table` definition that uses styles?

Reply:
```
def namespace_usage_panel(decimals, format, query, unit):
    return Table(
        columns=[
            Column(
                style=ColumnStyle(
                    decimals=decimals,
                    unit=format,
                ),
                text=unit,
                value='current',
            ),
        ],
        dataSource='prometheus',
        sort=ColumnSort(col=1, desc=True),
        span=6,
        styles=[
            ColumnStyle(
                alias='Namespace',
                pattern='Metric',
                type='String',
            ),
        ],
        targets=[
            Target(
                expr=query,
                intervalFactor=10,
                legendFormat='{{namespace}}',
                refId='A',
            ),
        ],
        title="%s by Namespace" % unit,
        transform=AGGREGATIONS_TRANSFORM,
    )


...some dashboard definition
        Row(
            collapse=True,
            panels=[
                namespace_usage_panel(
                    decimals=3,
                    format=SHORT_FORMAT,
                    query='namespace:kube_pod_container_resource_requests_cpu_cores:sum_onlyrunning',
                    unit='CPU Cores'
                ),
                namespace_usage_panel(
                    decimals=0,
                    format=BYTES_FORMAT,
                    query='namespace:kube_pod_container_resource_requests_memory_bytes:sum_onlyrunning',
                    unit='Memory'
                ),
            ],
            title='Per-namespace Usage',
        ),

```

Reply:
Grafana requires specification of the column styles by regex patterns on the column names.

In Python it's much easier in the common case to just specify the column style with the column, instead of having them separated.

Reply:
Thanks for the explanation!